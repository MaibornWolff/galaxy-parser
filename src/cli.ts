import fs from "node:fs/promises";
import yargs from "yargs";
import { GenericParser } from "./parser/GenericParser.js";
import { Configuration } from "./parser/Configuration.js";
import { updateNodeTypesMappingFile } from "./commands/import-grammars/ImportNodeTypes.js";
import { outputAsJson } from "./commands/outputMetrics.js";

export const parser = yargs()
    .command(
        "import-grammars",
        "(re-)import grammar expression types for supported languages",
        {},
        async () => {
            await updateNodeTypesMappingFile();
        },
    )
    .command(
        "parse [sources-path]",
        "parse file or folders recursively by given path and calculate metrics",
        (cmdYargs) => {
            return cmdYargs
                .positional("sources-path", {
                    describe: "path to sources",
                    type: "string",
                })
                .option("output-path", {
                    alias: "o",
                    type: "string",
                    description: "Output file path (required)",
                })
                .option("relative-paths", {
                    alias: "r",
                    type: "boolean",
                    default: false,
                    description:
                        "Write relative instead of absolute paths to the analyzed files in the output",
                })
                .option("exclusions", {
                    alias: "e",
                    type: "string",
                    description: "Exclude folders from scanning for files (comma separated list)",
                    default: "node_modules,.idea,dist,build,out,vendor",
                })
                .option("parse-h-as-c", {
                    alias: "hc",
                    type: "boolean",
                    description: "Parse all .h files as C instead of C++ (defaults to C++)",
                    default: false,
                })
                .option("parse-some-h-as-c", {
                    alias: "shc",
                    type: "string",
                    description:
                        "For the specified folders/files (comma separated list), parse .h files as C instead of C++. " +
                        "Ignored if parse-h-as-c is set.",
                    default: "",
                })
                .option("compress", {
                    alias: "c",
                    type: "boolean",
                    description: "output .gz-zipped file",
                    default: false,
                })
                .option("parse-dependencies", {
                    type: "boolean",
                    default: false,
                    description:
                        "EXPERIMENTAL: flag to enable dependency parsing (dependencies will be appended to the output file)",
                })
                .demandOption(["sources-path", "output-path"]);
        },
        async (argv) => {
            const configuration = new Configuration({
                /* eslint-disable @typescript-eslint/dot-notation */
                sourcesPath: await fs.realpath(argv["sources-path"]),
                outputPath: argv["output-path"],
                parseDependencies: argv["parse-dependencies"],
                exclusions: argv["exclusions"],
                parseAllHAsC: argv["parse-h-as-c"],
                parseSomeHAsC: argv["parse-some-h-as-c"],
                compress: argv["compress"],
                relativePaths: argv["relative-paths"],
                /* eslint-enable @typescript-eslint/dot-notation */
            });
            await parseSourceCode(configuration);
        },
    )
    .demandCommand()
    .strictCommands()
    .strictOptions();

async function parseSourceCode(configuration: Configuration): Promise<void> {
    try {
        console.time("Time to complete");

        const parser = new GenericParser(configuration);
        const results = await parser.calculateMetrics();

        console.log("#####################################");
        console.log("Metrics calculation finished.");
        console.timeEnd("Time to complete");

        outputAsJson(
            results.fileMetrics,
            results.unsupportedFiles,
            results.errorFiles,
            results.couplingMetrics,
            configuration.outputPath,
            configuration.compress,
        );
    } catch (error) {
        console.error("#####################################");
        console.error("#####################################");
        console.error("Metrics calculation failed with the following error:");
        console.error(error);
    }
}
