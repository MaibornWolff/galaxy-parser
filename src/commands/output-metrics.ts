import * as fs from "node:fs";
import { Readable } from "node:stream";
import zlib from "node:zlib";
import {
    type CouplingResult,
    type FileMetricResults,
    type MetricError,
} from "../parser/metrics/metric.js";
import { FileType } from "../parser/helper/language.js";

type OutputNode = {
    name: string;
    type: string;
    metrics: Record<string, number>;
};

type OutputInfoNode = {
    name: string;
    type: string;
    message: string;
};

type OutputRelationship = {
    from: string;
    to: string;
    metrics: {
        coupling: 100;
    };
};

/**
 * Writes the passed metrics into a json file.
 * @param fileMetrics Metrics calculated on single files.
 * @param unsupportedFiles List of files that cannot be analyzed.
 * @param errorFiles List of files that could not be parsed at all.
 * @param relationshipMetrics Relationship metrics.
 * @param outputFilePath Path to write the file to
 * @param compress Whether the file should be compressed
 */
export function outputAsJson(
    fileMetrics: Map<string, FileMetricResults>,
    unsupportedFiles: string[],
    errorFiles: string[],
    relationshipMetrics: CouplingResult,
    outputFilePath: string,
    compress: boolean,
): void {
    const output = buildOutputObject(
        fileMetrics,
        unsupportedFiles,
        errorFiles,
        relationshipMetrics,
    );
    const outputString = JSON.stringify(output).toString();

    if (compress) {
        outputFilePath = outputFilePath.endsWith(".gz") ? outputFilePath : outputFilePath + ".gz";
        dumpCompressed(outputString, outputFilePath);
    } else {
        fs.writeFileSync(outputFilePath, outputString);
    }

    console.log("Results saved to " + outputFilePath);
}

function buildOutputObject(
    fileMetrics: Map<string, FileMetricResults>,
    unknownFiles: string[],
    errorFiles: string[],
    relationshipMetrics: CouplingResult,
): {
    nodes: OutputNode[];
    info: OutputInfoNode[];
    relationships: OutputRelationship[];
} {
    const output: {
        nodes: OutputNode[];
        info: OutputInfoNode[];
        relationships: OutputRelationship[];
    } = {
        nodes: [],
        info: [],
        relationships: [],
    };

    const outputNodeReferenceLookUp = new Map<string, OutputNode>();

    const metricErrorsPerFile = new Map<string, Iterable<MetricError>>();

    for (const [filePath, fileMetricResults] of fileMetrics.entries()) {
        const metrics: Record<string, number> = {};
        if (fileMetricResults.metricErrors.length > 0) {
            metricErrorsPerFile.set(filePath, fileMetricResults.metricErrors);
        }

        for (const metricResult of fileMetricResults.metricResults) {
            metrics[metricResult.metricName] = metricResult.metricValue;
        }

        const outputNode: OutputNode = {
            name: filePath,
            type: fileMetricResults.fileType,
            metrics,
        };

        outputNodeReferenceLookUp.set(filePath, outputNode);
        output.nodes.push(outputNode);
    }

    // Info for unknown file types
    for (const filePath of unknownFiles) {
        const outputNode: OutputInfoNode = {
            name: filePath,
            type: FileType.Unsupported,
            message: "Unknown language or file extension",
        };

        output.info.push(outputNode);
    }

    // Info when parsing the syntax tree failed
    for (const filePath of errorFiles) {
        const outputNode: OutputInfoNode = {
            name: filePath,
            type: FileType.Error,
            message: "Error while parsing a syntax tree for the file",
        };

        output.info.push(outputNode);
    }

    // Info when the calculation of (some) metrics failed
    for (const [filePath, metricErrors] of metricErrorsPerFile) {
        let message = "Error while calculating the following metric(s) for the file:";
        for (const metricError of metricErrors) {
            message += " " + metricError.metricName;
        }

        const outputNode: OutputInfoNode = {
            name: filePath,
            type: FileType.Error,
            message,
        };

        output.info.push(outputNode);
    }

    // Merge relationship metrics with existing nodes or add new node
    const couplingMetrics = relationshipMetrics.metrics.entries();
    for (const [filePath, metricsMap] of couplingMetrics) {
        const existingOutputNode = outputNodeReferenceLookUp.get(filePath);
        if (existingOutputNode === undefined) {
            const newOutputNode: OutputNode = {
                name: filePath,
                type: FileType.SourceCode,
                metrics: { ...metricsMap },
            };
            output.nodes.push(newOutputNode);
        } else {
            existingOutputNode.metrics = { ...existingOutputNode.metrics, ...metricsMap };
        }
    }

    const couplingMetricResults = relationshipMetrics.relationships;

    for (const couplingMetricResult of couplingMetricResults) {
        output.relationships.push({
            from: couplingMetricResult.fromSource,
            to: couplingMetricResult.toSource,
            metrics: { coupling: 100 },
        });
    }

    return output;
}

function dumpCompressed(outputString: string, outputFilePath: string): void {
    const readableStream = new Readable();
    readableStream.push(outputString);

    const gzip = zlib.createGzip();
    const writeStream = fs.createWriteStream(outputFilePath);
    readableStream.pipe(gzip).pipe(writeStream);
}
