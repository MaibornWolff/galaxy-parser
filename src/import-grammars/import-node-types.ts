import fs from "node:fs/promises";
import { debuglog, type DebugLoggerFunction } from "node:util";
import { type NodeTypeConfig, NodeTypeCategory } from "../helper/model.js";
import nodeTypesConfig from "../parser/config/node-types-config.json" with { type: "json" };
import { type NodeType, type NodeTypes } from "../helper/node-types.js";
import { Language } from "../helper/language.js";
import { NodeTypesChangelog } from "./node-types-changelog.js";

let dlog: DebugLoggerFunction = debuglog("metric-gardener", (logger) => {
    dlog = logger;
});

/**
 * Maps all available languages (as they are used inside the node-types-config.json) to
 * the location of the node-types.json of the corresponding grammar.
 */
const languageToNodeTypeFiles = new Map<Language, string>([
    [Language.CSharp, "./node_modules/tree-sitter-c-sharp/src/node-types.json"],
    [Language.Go, "./node_modules/tree-sitter-go/src/node-types.json"],
    [Language.Java, "./node_modules/tree-sitter-java/src/node-types.json"],
    [Language.JavaScript, "./node_modules/tree-sitter-javascript/src/node-types.json"],
    [Language.Kotlin, "./node_modules/tree-sitter-kotlin/src/node-types.json"],
    [Language.PHP, "./node_modules/tree-sitter-php/php/src/node-types.json"],
    [Language.TypeScript, "./node_modules/tree-sitter-typescript/typescript/src/node-types.json"],
    [Language.TSX, "./node_modules/tree-sitter-typescript/tsx/src/node-types.json"],
    [Language.Python, "./node_modules/tree-sitter-python/src/node-types.json"],
    [Language.CPlusPlus, "./node_modules/tree-sitter-cpp/src/node-types.json"],
    [Language.Ruby, "./node_modules/tree-sitter-ruby/src/node-types.json"],
    [Language.Rust, "./node_modules/tree-sitter-rust/src/node-types.json"],
    [Language.Bash, "./node_modules/tree-sitter-bash/src/node-types.json"],
    [Language.C, "./node_modules/tree-sitter-c/src/node-types.json"],
    [Language.JSON, "./node_modules/tree-sitter-json/src/node-types.json"],
    [Language.YAML, "./node_modules/tree-sitter-yaml/src/node-types.json"],
]);

const pathToNodeTypesConfig = "./src/parser/config/node-types-config.json";

const nodeTypeMappings = new Map<string, NodeTypeConfig>();
const changelog = new NodeTypesChangelog();
const obsoleteDeactivatedLanguages = new Map<string, string[]>();

/**
 * Updates the node mappings for calculating metrics by importing the node-types.json files of the currently
 * installed language grammars.
 * Keeps the present mappings if the corresponding node type is still present in that language.
 * Removes all node types which are no longer present in the grammar.
 */
export async function updateNodeTypesMappingFile(): Promise<void> {
    nodeTypeMappings.clear();
    changelog.clear();

    try {
        const languageToNodeTypePromises = readNodeTypesJsons();
        const languageToPresentNodeTypes = importPresentNodeTypeMappings();

        for (const language of languageToNodeTypeFiles.keys()) {
            const presentNodeTypes = languageToPresentNodeTypes.get(language);
            const nodeTypesPromise = languageToNodeTypePromises.get(language);

            // Await reading node-types.json for the language:
            const nodeTypesJson = await nodeTypesPromise; // eslint-disable-line no-await-in-loop
            if (!nodeTypesJson) {
                throw new Error("No node-types.json found for language " + language);
            }

            if (nodeTypesJson instanceof Error) {
                throw nodeTypesJson;
            }

            updateLanguage(language, presentNodeTypes, nodeTypesJson);
        }

        // Make sure to write the change information before deleting any metric mapping.
        await changelog.writeChangelog(nodeTypeMappings);
        removeObsolete();
    } catch (error) {
        console.error("Error while updating the node mappings. Cancel update...");
        console.error(error);
        return;
    }

    try {
        await writeNewNodeTypeMappings();
    } catch (error) {
        console.error("Error while writing node-types-config.json");
        console.error(error);
    }
}

function readNodeTypesJsons(): Map<string, Promise<string | Error>> {
    const languageToNodeTypePromises = new Map<string, Promise<string | Error>>();

    for (const [language, fileLocation] of languageToNodeTypeFiles.entries()) {
        const json = fs.readFile(fileLocation, "utf8").catch((error: unknown) => {
            return new Error(
                `Error while reading a node-types.json file from ${fileLocation}:\n${String(error)}`,
            ); // To be handled when awaiting the result.
        });
        languageToNodeTypePromises.set(language, json);
    }

    return languageToNodeTypePromises;
}

function importPresentNodeTypeMappings(): Map<string, Set<string>> {
    const presentNodeTypesForLanguage = new Map<string, Set<string>>();
    for (const language of languageToNodeTypeFiles.keys()) {
        presentNodeTypesForLanguage.set(language, new Set());
    }

    const allNodeTypes = nodeTypesConfig as NodeTypeConfig[];

    for (const nodeType of allNodeTypes) {
        nodeTypeMappings.set(nodeType.type_name, nodeType);

        for (const language of nodeType.languages) {
            presentNodeTypesForLanguage.get(language)?.add(nodeType.type_name);
        }
    }

    return presentNodeTypesForLanguage;
}

function updateLanguage(
    language: Language,
    presentNodes: Set<string> | undefined,
    nodeTypesJson: string,
): void {
    const grammarNodeTypes = JSON.parse(nodeTypesJson) as NodeTypes;

    const toRemove = new Set<string>(presentNodes);
    for (const grammarNodeType of grammarNodeTypes) {
        // Ignore all unnamed syntax nodes that are no binary expressions (as we require them) to get a "kind of"
        // abstract syntax tree by removing syntax details.
        // See also https://tree-sitter.github.io/tree-sitter/using-parsers#named-vs-anonymous-nodes
        if (!grammarNodeType.named) {
            dlog(`Excluded unnamed node type ${grammarNodeType.type} for language ${language}`);
            continue;
        }

        if (grammarNodeType.type === "binary_expression") {
            updateBinaryExpressions(language, grammarNodeType, toRemove);
            continue;
        }

        if (grammarNodeType.subtypes === undefined) {
            toRemove.delete(grammarNodeType.type);
            updateOrAddExpression(language, grammarNodeType.type);
        } else {
            // Do only include the (concrete) subtypes of a supertype node (types that have subtypes),
            // as supertypes represent abstract categories of syntax nodes
            // (see https://tree-sitter.github.io/tree-sitter/using-parsers).
            for (const subNodeType of grammarNodeType.subtypes) {
                toRemove.delete(subNodeType.type);
                updateOrAddExpression(language, subNodeType.type);
            }
        }
    }

    removeNodeTypesForLanguage(language, toRemove);
}

function updateBinaryExpressions(
    language: Language,
    grammarNodeType: NodeType,
    removedNodeTypes: Set<string>,
): void {
    if (grammarNodeType.fields?.operator?.types !== undefined) {
        for (const binaryOperatorType of grammarNodeType.fields.operator.types) {
            const { type: binaryOperator } = binaryOperatorType;
            const mapKey = grammarNodeType.type + "_" + binaryOperator;
            removedNodeTypes.delete(mapKey);

            updateOrAddExpression(
                language,
                mapKey,
                NodeTypeCategory.OtherBinaryExpression, // Can be adjusted manually in the node-types-config later
                grammarNodeType.type,
                binaryOperator,
            );
        }
    }
}

/**
 * Adds the node type to the specified language.
 * Either updates an existing mapping for that node type or adds a new node type to the mappings.
 * @param language Language to add this node type to.
 * @param nodeTypeName Name of the node type.
 * @param category Category of the node type to use if a new node type mapping is created.
 * @param grammarNodeTypeName Name of the node type in the grammar, if different to the nodeTypeName.
 * @param operator Operator of the node type to use if a new node type mapping is created.
 */
function updateOrAddExpression(
    language: Language,
    nodeTypeName: string,
    category: NodeTypeCategory = NodeTypeCategory.Other,
    grammarNodeTypeName?: string,
    operator?: string,
): void {
    const nodeType = nodeTypeMappings.get(nodeTypeName);

    if (nodeType === undefined) {
        nodeTypeMappings.set(nodeTypeName, {
            type_name: nodeTypeName,
            category,
            languages: [language],
            grammar_type_name: grammarNodeTypeName,
            operator,
        });
        changelog.addedNewNode(nodeTypeName, language);
        dlog('New node type "' + nodeTypeName + '" was added for language "' + language + '".');
    } else if (!nodeType.languages.includes(language)) {
        changelog.addedNodeToLanguage(nodeType, language);
        nodeType.languages.push(language);
        dlog('Language "' + language + '" was added to node type "' + nodeTypeName + '".');
    }
}

function removeNodeTypesForLanguage(language: Language, removedNodeTypes: Set<string>): void {
    for (const [nodeTypeName, nodeType] of nodeTypeMappings) {
        if (removedNodeTypes.has(nodeTypeName)) {
            const index = nodeType.languages.indexOf(language);
            nodeType.languages.splice(index, 1);

            changelog.removedNodeFromLanguage(nodeType, language);
            dlog('Node type "' + nodeTypeName + '" was removed for language "' + language + '".');
            if (
                nodeType.category !== NodeTypeCategory.Other &&
                !nodeType.deactivated_for_languages?.includes(language)
            ) {
                console.warn(
                    '## Attention required! Language "' +
                        language +
                        '" no longer includes the node type "' +
                        nodeTypeName +
                        '", which was used for calculating metric(s) under the category ' +
                        nodeType.category +
                        ". You may have to add a new node of that language to this category in node-types-config.json. ##",
                );
            }

            if (nodeType.deactivated_for_languages?.includes(language)) {
                const obsolete = obsoleteDeactivatedLanguages.get(nodeTypeName) ?? [];
                obsolete.push(language);
                obsoleteDeactivatedLanguages.set(nodeTypeName, obsolete);
            }
        }
    }
}

function removeObsolete(): void {
    for (const [nodeTypeName, nodeType] of nodeTypeMappings) {
        if (nodeType.languages.length === 0) {
            if (nodeType.category !== NodeTypeCategory.Other) {
                console.warn(
                    '#### Intervention required! Removing the node type "' +
                        nodeTypeName +
                        '", which was used for calculating metric(s) under the category ' +
                        nodeType.category +
                        ". You may have to add a new node to the metric(s) in node-types-config.json. ####",
                );
            }

            nodeTypeMappings.delete(nodeTypeName);
            dlog(
                `Removed node type ${nodeTypeName} as it is no longer used in any language grammar`,
            );
        } else {
            const obsolete = obsoleteDeactivatedLanguages.get(nodeTypeName);
            nodeType.deactivated_for_languages = nodeType.deactivated_for_languages?.filter(
                (lang) => !obsolete?.includes(lang),
            );
            if (nodeType.deactivated_for_languages?.length === 0) {
                delete nodeType.deactivated_for_languages;
            }
        }
    }
}

async function writeNewNodeTypeMappings(): Promise<void> {
    // Save the updated mappings:
    await fs.writeFile(
        pathToNodeTypesConfig,
        JSON.stringify([...nodeTypeMappings.values()], null, 4),
    );
    console.log("####################################");
    console.log(
        'Successfully updated node type mappings. File saved to "' + pathToNodeTypesConfig + '".',
    );
    console.log("####################################");
}
