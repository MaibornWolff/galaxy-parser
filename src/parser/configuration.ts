/**
 * Parameters of the constructor of {@link Configuration}.
 * Represents configuration options that can be provided by the user via command line arguments.
 */
export type ConfigurationParameters = {
    /**
     * Path to the source files to be parsed.
     */
    sourcesPath: string;
    /**
     * Path where the output file with the calculated metrics should be stored.
     */
    outputPath: string;
    /**
     * Whether dependencies should be analyzed.
     */
    parseDependencies: boolean;
    /**
     * Folders to exclude from being searched for files to be parsed.
     */
    exclusions: string;
    /**
     * Whether to parse all .h files as C instead of C++.
     */
    parseAllHAsC: boolean;
    /**
     * Folders or file names for which .h files should be parsed as C instead of C++ files. Ignored if parseAllAsC is set.
     */
    parseSomeHAsC: string;
    /**
     * Whether to compress the output file with the calculated metrics in a zip archive.
     */
    compress: boolean;
    /**
     * Whether to include the relative file paths or absolute paths of the analyzed files in the output.
     */
    relativePaths: boolean;
};

/**
 * Configures the files to be parsed and the metrics to be calculated.
 */
export class Configuration {
    /**
     * Resolved, absolute path to the source files to be parsed.
     */
    readonly sourcesPath: string;

    /**
     * Path where the output file with the calculated metrics should be stored.
     */
    readonly outputPath: string;

    /**
     * Whether dependencies should be analyzed.
     */
    readonly parseDependencies: boolean;

    /**
     * Folders to exclude from being searched for files to be parsed.
     */
    readonly exclusions: Set<string>;

    /**
     * Whether to parse all .h files as C instead of C++.
     */
    readonly parseAllHAsC: boolean;

    /**
     * Folders or file names for which .h files should be parsed as C instead of C++ files.
     */
    readonly parseSomeHAsC: Set<string>;

    /**
     * Whether to compress the output file with the calculated metrics in a zip archive.
     */
    readonly compress: boolean;

    /**
     * Whether to include the relative file paths or absolute paths of the analyzed files in the output.
     */
    readonly relativePaths: boolean;

    /**
     * Constructs a new {@link Configuration} object by specifying the configuration options passed by the user
     * as command line arguments.
     * @param parameters {@link Parameters} object containing the configuration options.
     */
    constructor(parameters: ConfigurationParameters) {
        this.sourcesPath = parameters.sourcesPath;
        this.outputPath = parameters.outputPath;
        this.parseDependencies = parameters.parseDependencies;

        this.exclusions = new Set(
            parameters.exclusions.length > 0
                ? parameters.exclusions.split(",").map((exclusion) => exclusion.trim())
                : null,
        );

        this.parseAllHAsC = parameters.parseAllHAsC;

        this.parseSomeHAsC = new Set(
            !parameters.parseAllHAsC && parameters.parseSomeHAsC.length > 0
                ? parameters.parseSomeHAsC
                      .split(",")
                      .map((folderOrFileName) => folderOrFileName.trim())
                : null,
        );

        this.compress = parameters.compress;
        this.relativePaths = parameters.relativePaths;
    }
}
