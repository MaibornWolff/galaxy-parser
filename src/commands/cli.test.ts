import fs from "node:fs/promises";
import { afterAll, describe, expect, it, vi } from "vitest";
import { mockConsole } from "../../test/metric-end-results/test-helper.js";
import * as ImportNodeTypes from "../import-grammars/import-node-types.js";
import { Configuration } from "../parser/configuration.js";
import { type CouplingResult, type FileMetricResults } from "../parser/metrics/metric.js";
import { parser } from "./cli.js";
import * as outputMetrics from "./output-metrics.js";

const parserConstructor = vi.hoisted(() => vi.fn<[Configuration]>());
const parserCalculateMetrics = vi.hoisted(() =>
    vi.fn<
        [], // eslint-disable-line @typescript-eslint/ban-types
        Promise<{
            couplingMetrics: CouplingResult;
            fileMetrics: Map<string, FileMetricResults>;
            unsupportedFiles: string[];
            errorFiles: string[];
        }>
    >(),
);
vi.mock("../parser/generic-parser.js", () => ({
    GenericParser: class GenericParser {
        calculateMetrics = parserCalculateMetrics;
        constructor(config: Configuration) {
            parserConstructor(config);
        }
    },
}));

describe("cli", () => {
    afterAll(() => {
        vi.resetModules();
    });

    itShouldOfferHelp();

    describe("parse command", () => {
        const expectedConfig = new Configuration({
            sourcesPath: ".",
            outputPath: "metrics.json",
            parseDependencies: false,
            exclusions: "node_modules,.idea,dist,build,out,vendor",
            parseAllHAsC: false,
            parseSomeHAsC: "",
            compress: false,
            relativePaths: false,
        });
        const expectedMetrics = {
            couplingMetrics: { relationships: [], metrics: new Map() },
            fileMetrics: new Map(),
            unsupportedFiles: ["unsupported"],
            errorFiles: ["error"],
        };

        itShouldOfferHelp("parse");

        it("should calculate and output metrics", async () => {
            mockConsole();
            vi.spyOn(fs, "realpath").mockImplementation(async (path) => path.toString());
            parserCalculateMetrics.mockResolvedValue(expectedMetrics);
            vi.spyOn(outputMetrics, "outputAsJson").mockReset();

            await parser.parse("parse . -o metrics.json");

            expect(console.time).toHaveBeenCalledWith("Time to complete");
            expect(parserConstructor).toHaveBeenCalledWith(expectedConfig);
            expect(parserCalculateMetrics).toHaveBeenCalled();
            expect(console.log).toHaveBeenCalledWith("#####################################");
            expect(console.log).toHaveBeenCalledWith("Metrics calculation finished.");
            expect(console.timeEnd).toHaveBeenCalledWith("Time to complete");
            expect(outputMetrics.outputAsJson).toHaveBeenCalledWith({
                fileMetrics: expectedMetrics.fileMetrics,
                unsupportedFiles: expectedMetrics.unsupportedFiles,
                errorFiles: expectedMetrics.errorFiles,
                relationshipMetrics: expectedMetrics.couplingMetrics,
                outputFilePath: expectedConfig.outputPath,
                compress: expectedConfig.compress,
            });
        });

        it("should configure the parser with the correct options", async () => {
            mockConsole();
            vi.spyOn(fs, "realpath").mockImplementation(async (path) => path.toString());
            parserCalculateMetrics.mockResolvedValue(expectedMetrics);
            vi.spyOn(outputMetrics, "outputAsJson").mockReset();

            await parser.parse("parse . -o metrics.json -r");
            expect(parserConstructor).toHaveBeenNthCalledWith(1, {
                ...expectedConfig,
                relativePaths: true,
            });

            await parser.parse("parse . -o metrics.json -e excluded");
            expect(parserConstructor).toHaveBeenNthCalledWith(2, {
                ...expectedConfig,
                exclusions: new Set(["excluded"]),
            });

            await parser.parse("parse . -o metrics.json --hc");
            expect(parserConstructor).toHaveBeenNthCalledWith(3, {
                ...expectedConfig,
                parseAllHAsC: true,
            });

            await parser.parse("parse . -o metrics.json --shc some");
            expect(parserConstructor).toHaveBeenNthCalledWith(4, {
                ...expectedConfig,
                parseSomeHAsC: new Set(["some"]),
            });

            await parser.parse("parse . -o metrics.json -c");
            expect(parserConstructor).toHaveBeenNthCalledWith(5, {
                ...expectedConfig,
                compress: true,
            });

            await parser.parse("parse . -o metrics.json --parse-dependencies");
            expect(parserConstructor).toHaveBeenNthCalledWith(6, {
                ...expectedConfig,
                parseDependencies: true,
            });
        });

        it("should log error if metrics calculation fails", async () => {
            mockConsole();
            const error = new Error("Error");
            parserCalculateMetrics.mockRejectedValue(error);
            vi.spyOn(outputMetrics, "outputAsJson").mockReset();

            await parser.parse("parse . -o metrics.json");

            expect(console.error).toHaveBeenCalledTimes(4);
            expect(console.error).toHaveBeenCalledWith("#####################################");
            expect(console.error).toHaveBeenCalledWith("#####################################");
            expect(console.error).toHaveBeenCalledWith(
                "Metrics calculation failed with the following error:",
            );
            expect(console.error).toHaveBeenCalledWith(error);
            expect(outputMetrics.outputAsJson).not.toHaveBeenCalled();
        });

        it("should log error if output of metrics fails", async () => {
            mockConsole();
            parserCalculateMetrics.mockResolvedValue(expectedMetrics);
            const error = new Error("Error");
            vi.spyOn(outputMetrics, "outputAsJson").mockImplementation(() => {
                throw error;
            });

            await parser.parse("parse . -o metrics.json");

            expect(console.error).toHaveBeenCalledTimes(4);
            expect(console.error).toHaveBeenCalledWith("#####################################");
            expect(console.error).toHaveBeenCalledWith("#####################################");
            expect(console.error).toHaveBeenCalledWith(
                "Metrics calculation failed with the following error:",
            );
            expect(console.error).toHaveBeenCalledWith(error);
        });
    });

    function itShouldOfferHelp(command = ""): void {
        it("should offer help", async () => {
            mockConsole();
            await parser.exitProcess(false).parse(`${command} --help`);
            expect(vi.mocked(console.log).mock.lastCall?.[0]).toMatchSnapshot();
        });
    }
});
