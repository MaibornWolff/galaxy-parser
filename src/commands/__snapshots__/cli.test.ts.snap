// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`cli > parse command > should offer help 1`] = `
"process.js parse [sources-path]

parse file or folders recursively by given path and calculate metrics

Positionals:
  sources-path  path to sources                              [string] [required]

Options:
      --help                      Show help                            [boolean]
      --version                   Show version number                  [boolean]
  -o, --output-path               Output file path (required)[string] [required]
  -r, --relative-paths            Write relative instead of absolute paths to th
                                  e analyzed files in the output
                                                      [boolean] [default: false]
  -e, --exclusions                Exclude folders from scanning for files (comma
                                   separated list)
                  [string] [default: "node_modules,.idea,dist,build,out,vendor"]
      --parse-h-as-c, --hc        Parse all .h files as C instead of C++ (defaul
                                  ts to C++)          [boolean] [default: false]
      --parse-some-h-as-c, --shc  For the specified folders/files (comma separat
                                  ed list), parse .h files as C instead of C++.
                                  Ignored if parse-h-as-c is set.
                                                          [string] [default: ""]
  -c, --compress                  output .gz-zipped file
                                                      [boolean] [default: false]
      --parse-dependencies        EXPERIMENTAL: flag to enable dependency parsin
                                  g (dependencies will be appended to the output
                                   file)              [boolean] [default: false]"
`;

exports[`cli > should offer help 1`] = `
"process.js <command>

Commands:
  process.js parse [sources-path]  parse file or folders recursively by given pa
                                   th and calculate metrics

Options:
  --help     Show help                                                 [boolean]
  --version  Show version number                                       [boolean]"
`;
